sdk/internal/perf/perf.go

```diff
@@ -0,0 +1,329 @@
+func runGlobalCleanup(p PerfTest) error {
+	err := p.GlobalCleanup(context.Background())
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+type runResult struct {
+	count         int
+	timeInSeconds float64
+	err           error
+}
+
+func runPerfTest(p NewPerfTest) error {
+	err := runGlobalSetup(p(nil))
+	if err != nil {
+		panic(err)
+	}
+
+	var channels []chan runResult
+	var perfTests []PerfTest
+
+	fmt.Println("=== Test ===")
+
+	w := tabwriter.NewWriter(os.Stdout, 16, 8, 1, ' ', tabwriter.AlignRight)
+	fmt.Fprintln(w, "Current\tTotal\tAverage\t")
+	err = w.Flush()
+	if err != nil {
+		panic(err)
+	}
+
+	for idx := 0; idx < Parallel; idx++ {
+		options := &PerfTestOptions{}
+		if TestProxy != "" {
+			transporter, err := NewProxyTransport(&TransportOptions{TestName: p(nil).GetMetadata().Name})
+			if err != nil {
+				panic(err)
+			}
+			options.ProxyInstance = transporter
+		} else {
+			options.ProxyInstance = defaultHTTPClient
+		}
+		options.ParallelIndex = idx
+
+		perfTest := p(options)
+		perfTests = append(perfTests, perfTest)
+
+		// Run the setup for a single instance
+		err := runSetup(perfTest)
+		if err != nil {
+			return err
+		}
+
+		// Create a thread for running a single test
+		wg.Add(1)
+		c := make(chan runResult, 1) // Create a buffered channel
+		channels = append(channels, c)
+		go runTest(perfTest, c)
+	}
+
+	wg.Wait()
+
+	opsPerSecond := 0.0
+	totalOperations := 0
+	// Get the results from the channels
+	for _, channel := range channels {
+		result := <-channel
+		if err != nil {
+			panic(err)
+		}
+
+		opsPerSecond += float64(result.count) / result.timeInSeconds
+		totalOperations += result.count
+	}
+
+	// Run Cleanup on each parallel instance
+	for _, pTest := range perfTests {
+		err = runCleanup(pTest)
+		if err != nil {
+			panic(err)
+		}
+	}
+
+	err = runGlobalCleanup(p(nil))
+
+	fmt.Println("\n=== Results ===")
+	secondsPerOp := 1.0 / opsPerSecond
+	weightedAvgSec := float64(totalOperations) / opsPerSecond
+	fmt.Printf(
+		"Completed %s operations in a weighted-average of %.2fs (%s ops/s, %.3f s/op)",
+		commaIze(totalOperations),
+		weightedAvgSec,
+		commaIze(int(opsPerSecond)),
+		secondsPerOp,
+	)
+
+	return err
+}
+
+// testsToRun trims the slice of PerfTest to only those that are flagged as running.
+func testsToRun(registered []NewPerfTest) NewPerfTest {
+	args := os.Args[1:]
+	for _, r := range registered {
+		p := r(nil)
+		for _, arg := range args {
+			if p.GetMetadata().Name == arg {
+				return r
+			}
+		}
+	}
+
+	return nil
+}
+
+var registerCalled bool
+
+// RegisterArguments is used to
+func RegisterArguments(f func()) {
+	f()
+	registerCalled = true
+}
+
+// Run runs all individual tests
+func Run(perfTests []NewPerfTest) {
+	// Start with adding all of our arguments
+	pflag.IntVarP(&Duration, "duration", "d", 10, "Duration of the test in seconds. Default is 10.")
+	pflag.StringVarP(&TestProxy, "test-proxies", "x", "", "whether to target http or https proxy (default is neither)")
```

Comment:
The `test-proxies` parameter needs to accept a list of URIs separated by `;`.  The proxy URLs should be assigned round-robin to the parallel test instances.  So for `--parallel 5 --test-proxies http://proxy1;http://proxy2`, test instances 1, 3, and 5 should be assigned proxy1, and test instances 2 and 4 should be assigned proxy2.  This feature is used for scenarios where multiple test proxies are required to bottleneck the client.

https://github.com/Azure/azure-sdk-for-python/blob/main/tools/azure-devtools/src/azure_devtools/perfstress_tests/_perf_stress_runner.py#L106

Reply:
Resolved